chore: add README to clock testing (#13583)
coder/coder: clock/README.md

# Quartz
A Go time testing library for writing deterministic unit tests

_Note: Quartz is the name I'm targeting for the standalone open source project when we spin this
out._

## Why another time testing library?

> Writing good unit tests for components and functions that use the `time` package is difficult, and
> several open source libraries exist.  In building Quartz, we took some inspiration from

Reviewer: ```suggestion
Writing good unit tests for components and functions that use the `time` package is difficult, even though
several open source libraries exist.  In building Quartz, we took some inspiration from
```

 * [github.com/benbjohnson/clock](https://github.com/benbjohnson/clock)
 * Tailscale's [tstest.Clock](https://github.com/coder/tailscale/blob/main/tstest/clock.go)
 * [github.com/aspenmesh/tock](https://github.com/aspenmesh/tock)

Quartz shares the high level design of a `Clock` interface that closely resembles the functions in

> the `time` standard library, and a "real" clock passes thru to the standard libary in production,

Reviewer: ```suggestion
the `time` standard library, and a "real" clock passes through to the standard library in production,
```

Author: https://www.youtube.com/watch?v=FXBHY7uco0Y

while a mock clock gives precise control in testing.

Our high level goal is to write unit tests that

1. execute quickly

> 2. don't flake

Reviewer: suggestion: avoid colloquialism
```suggestion
2. execute reliably
```

Author: While arguably "jargon", I think flaky is more precise. It refers to tests that succeed or fail due to uncontrolled randomness.

A test might not be reliable for other reasons, such as not actually testing or asserting the thing is purports to.

3. are straightforward to write and understand

For several reasons, this is a tall order when it comes to code that depends on time, and we found
the existing libraries insufficient for our goals.

> For tests to execute quickly without flakes, we want to focus on _determinism_: the test should run

Reviewer: ```suggestion
For tests to execute quickly and reliably, we want to focus on _determinism_: the test should run
```

In our experience, an extremely common use case is creating a ticker then doing a 2-arm `select`
with ticks in one and context expiring in another, i.e.

```go
t := time.NewTicker(duration)
for {
    select {
    case <-ctx.Done():
        return ctx.Err()
    case <-t.C:
        err := do()
        if err != nil {
            return err
        }
    }
}
```

In Quartz, we refactor this to be more compact and testing friendly:

```go
t := clock.TickerFunc(ctx, duration, do)
return t.Wait()
```

This affords the mock `Clock` the ability to explicitly know when processing of a tick is finished
because it's wrapped in the function passed to `TickerFunc` (`do()` in this example).

In Quartz, when you advance the clock, you are returned an object you can `Wait()` on to ensure all
ticks and timers triggered are finished. This solves the first race condition in the example.

(As an aside, we still support a traditional standard library-style `Ticker`, if you want to keep
your code as close as possible to the standard library, or if you need to use the channel in a
larger `select` block, but in that case you'll have to find some other mechanism to sync tick

> processing to your test code.)

Reviewer: Run-on sentence, consider breaking up.
